File: python/bench.py
import mysql.connector
from threading import Thread
import time
import random


HOST = '127.0.0.1'
PORT = '4000'
USER = 'root'
PASSWORD = ''
DATABASE = 'test'

NUM_TABLES = 10
SIZE_TABLE = 100000


def point_get(cursor):
    random_id = random.randint(0, SIZE_TABLE - 1)
    random_table = random.randint(1, NUM_TABLES)
    query = f"SELECT * FROM sbtest{random_table} WHERE id = {random_id}"
    cursor.execute(query)
    result = cursor.fetchall()
    return

def simple_range(cursor):
    random_low = random.randint(0, SIZE_TABLE - 1)
    random_up = random.randint(random_low, SIZE_TABLE - 1)
    random_table = random.randint(1, NUM_TABLES)
    query = f"SELECT c FROM sbtest{random_table} WHERE id between {random_low} and {random_up}"
    cursor.execute(query)
    result = cursor.fetchall()
    return

def sum_ranges(cursor):
    random_low = random.randint(0, SIZE_TABLE - 1)
    random_up = random.randint(random_low, SIZE_TABLE - 1)
    random_table = random.randint(1, NUM_TABLES)
    query = f"SELECT SUM(k) FROM sbtest{random_table} WHERE id BETWEEN {random_low} AND {random_up}"
    cursor.execute(query)
    result = cursor.fetchall()
    return

def order_ranges(cursor):
    random_low = random.randint(0, SIZE_TABLE - 1)
    random_up = random.randint(random_low, SIZE_TABLE - 1)
    random_table = random.randint(1, NUM_TABLES)
    query = f"SELECT c FROM sbtest{random_table} WHERE id BETWEEN {random_low} AND {random_up} ORDER BY c"
    cursor.execute(query)
    result = cursor.fetchall()
    return

def distinct_ranges(cursor):
    random_low = random.randint(0, SIZE_TABLE - 1)
    random_up = random.randint(random_low, SIZE_TABLE - 1)
    random_table = random.randint(1, NUM_TABLES)
    query = f"SELECT DISTINCT c FROM sbtest{random_table} WHERE id BETWEEN {random_low} AND {random_up} ORDER BY c"
    cursor.execute(query)
    result = cursor.fetchall()
    return


### Sysbench Write Pattern
# non_index_updates = "UPDATE sbtest%u SET c=? WHERE id=?"
# deletes = "DELETE FROM sbtest%u WHERE id=?"
###

SQLS = [point_get]


# Function to execute queries
def execute_queries(thread_id):
    conn = mysql.connector.connect(
        host=HOST,
        port=PORT,
        user=USER,
        password=PASSWORD,
        database=DATABASE
    )

    # Connect to MySQL database
    cursor = conn.cursor()

    interval = 5 # in sec

    # Execute sysbench-like queries
    start_time = time.time()
    try:
        while True:
            time.sleep(interval)
            random.choice(SQLS)(cursor)
    except Exception as e:
        print(e)

    end_time = time.time()

    # Print execution time for each thread
    print(f"Thread {thread_id} finished in {end_time - start_time} seconds")

    # Conn not closed
    cursor.close()
    return


# Function for idle connection
def idle(thread_id):
    conn = mysql.connector.connect(
        host=HOST,
        port=PORT,
        user=USER,
        password=PASSWORD,
        database=DATABASE
    )

    # Execute sysbench-like queries
    start_time = time.time()
    try:
        while True:
            time.sleep(500)
    except Exception as e:
        print(e)

    end_time = time.time()

    # Print execution time for each thread
    print(f"Thread {thread_id} finished in {end_time - start_time} seconds")
    return


if __name__ == "__main__":
    # Number of threads (connections)
    tasks = [idle] * 100 + [execute_queries] * 10

    # Start threads
    threads = []
    for i, t in enumerate(tasks):
        thread = Thread(target=t, args=(i,))
        thread.start()
        threads.append(thread)

    # Wait for all threads to finish
    try:
        for thread in threads:
            thread.join()
    except KeyboardInterrupt:
        print("Done")
        import sys
        sys.exit(0)

File: golang/bench.go
package main

import (
	"context"
	"database/sql"
	"fmt"
	"math"
	"math/rand"
	"sync"
	"sync/atomic"
	"time"

	_ "github.com/go-sql-driver/mysql"
)

const (
	host         = "10.2.8.2"
	port         = "4501"
	user         = "root"
	password     = ""
	database     = "test"
	numTables    = 10
	sizeTable    = 100000
	numThreads   = 11000
	numIdle      = 10000
	numQueries   = 10
	queryPeriod  = 5 * time.Second
	reportPeriod = 10 * time.Second
)

var queries = []func(*sql.DB, *sync.WaitGroup){
	pointGet,
}

var connectionCount int32

func pointGet(db *sql.DB, wg *sync.WaitGroup) {
	randomID := rand.Intn(sizeTable)
	randomTable := rand.Intn(numTables) + 1
	query := fmt.Sprintf("SELECT * FROM sbtest%d WHERE id = ?", randomTable)
	db.Query(query, randomID)
}

func executeQueries(db *sql.DB, wg *sync.WaitGroup) {
	defer wg.Done()

	for {
		queryFunc := queries[rand.Intn(len(queries))]
		queryFunc(db, wg)
	}
}

func idle(db *sql.DB, wg *sync.WaitGroup) {
	defer wg.Done()
	for {
		time.Sleep(1 * time.Second)
		x := rand.Intn(4839280)
		db.Exec(fmt.Sprintf("select %d", x))
	}
}

func withCount(ctx context.Context, db *sql.DB, wg *sync.WaitGroup, task func(db *sql.DB, wg *sync.WaitGroup)) {
	atomic.AddInt32(&connectionCount, 1)
	defer atomic.AddInt32(&connectionCount, -1)
	task(db, wg)
}

func withDB(ctx context.Context, wg *sync.WaitGroup, task func(db *sql.DB, wg *sync.WaitGroup)) {
	atomic.AddInt32(&connectionCount, 1)
	defer atomic.AddInt32(&connectionCount, -1)
	db, err := sql.Open("mysql", fmt.Sprintf("%s:%s@tcp(%s:%s)/%s", user, password, host, port, database))
	db.SetMaxIdleConns(math.MaxInt32)
	db.SetMaxOpenConns(math.MaxInt32)
	db.SetConnMaxIdleTime(-1)
	db.SetConnMaxLifetime(-1)
	if err != nil {
		fmt.Println("Error connecting to database:", err)
		return
	}
	defer db.Close()
	task(db, wg)
}

func reportConnections() {
	for {
		time.Sleep(reportPeriod)
		count := atomic.LoadInt32(&connectionCount)
		fmt.Printf("Number of running connections: %d\n", count)
	}
}

func main() {
	db, err := sql.Open("mysql", fmt.Sprintf("%s:%s@tcp(%s:%s)/%s", user, password, host, port, database))
	if err != nil {
		fmt.Println("Error connecting to database:", err)
		return
	}
	defer db.Close()
	db.SetMaxIdleConns(math.MaxInt32)
	db.SetMaxOpenConns(math.MaxInt32)

	var wg sync.WaitGroup

	// Start idle connections
	for i := 0; i < numIdle; i++ {
		wg.Add(1)
		go withCount(context.Background(), db, &wg, idle)
	}

	// Start threads for executing queries
	for i := 0; i < numThreads-numIdle; i++ {
		wg.Add(1)
		go withCount(context.Background(), db, &wg, executeQueries)
	}

	// Start goroutine to report connections
	go reportConnections()

	// Wait for all threads to finish
	wg.Wait()
}
